openhcl/page_pool_alloc/src/lib.rs

```diff
@@ -0,0 +1,494 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+//! This module implements a page memory allocator for allocating pages from a
+//! given portion of the guest address space.
+
+#![cfg(unix)]
+#![warn(missing_docs)]
+
+mod device_dma;
+
+pub use device_dma::PagePoolDmaBuffer;
+
+#[cfg(all(feature = "vfio", target_os = "linux"))]
+use anyhow::Context;
+#[cfg(all(feature = "vfio", target_os = "linux"))]
+use hcl::ioctl::MshvVtlLow;
+use hvdef::HV_PAGE_SIZE;
+use inspect::Inspect;
+use parking_lot::Mutex;
+use std::num::NonZeroU64;
+use std::sync::Arc;
+use thiserror::Error;
+use vm_topology::memory::MemoryRangeWithNode;
+
+/// Error returned when unable to allocate memory.
+#[derive(Debug, Error)]
+#[error("unable to allocate page pool size {size} with tag {tag}")]
+pub struct PagePoolOutOfMemory {
+    size: u64,
+    tag: String,
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+enum State {
+    Free {
+        base_pfn: u64,
+        pfn_bias: u64,
+        size_pages: u64,
+    },
+    Allocated {
+        base_pfn: u64,
+        pfn_bias: u64,
+        size_pages: u64,
+        /// This is an index into the outer [`PagePoolInner`]'s device_ids
+        /// vector.
+        device_id: usize,
+        tag: String,
+    },
+}
+
+#[derive(Inspect, Debug, Clone, Copy, PartialEq, Eq)]
+enum PoolType {
```

Comment:
Probably we should think about how to split the core pool algorithms out from the HCL-specific memory handling. But this is fine for now.